### PR TITLE
RetroPlayer: Merge savestate metadata into .sav file

### PR DESCRIPTION
**Pull request by garbear**
## Description

Introduced in https://github.com/xbmc/xbmc/pull/12434, RetroPlayer uses two files for autosave: a .sav file for the memory data, and an .xml file for metadata. This combines both memory data and metadata into a single .sav file:

```
Chrono Trigger (U) [!].smc   (the ROM)
Chrono Trigger (U) [!].sav   (the state and metadata)
```

In this PR, FlatBuffers is used to serialize metadata. This library was chosen for performance reasons, as it is the only popular serialization library that allows for zero-copy. Although it requires a native compiler, the target library is headers-only.

Backwards compatibility with XML files was not retained due to the extra code it would take, and RetroPlayer is still alpha software.

As a result of this PR, savestates are several hundred kb smaller, half as many files, and loading/saving should be faster on slow devices due to skipping costly XML operations.

I spent a LONG time on the build system stuff, and I'm not sure if I did everything correctly.

## Motivation and Context
First abstraction for the new Saved Games manager.

## How Has This Been Tested?
Tested on Windows 7 and OSX. 

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Cosmetic change (non-breaking change that doesn't touch code)
- [ ] None of the above (please explain below)

### CODE REVIEW

**Comment by garbear** [comment]
TODO: The savestate needs a version field

(testing depends) jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

File: xbmc/cores/RetroPlayer/messages/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+set(MESSAGES savestate.fbs
+)
+
+core_add_messages(retroplayer_messages)
+set(EXTRA_INCLUDES ${EXTRA_INCLUDES} PARENT_SCOPE)
```
Comment (garbear) [inline xbmc/cores/RetroPlayer/messages/CMakeLists.txt]: @Rechi I need to pass the generated header's directory to the root cmake file. Is this an OK way to do that?

**Review by wsnipex — CHANGES_REQUESTED** [review]
cmake stuff doesn't conform to our standards and needs serious changes.
- move all flatbuffer building to depends
- fix the findmodule to allow bulding with external and internal flatbuffer and just use FLATBUFFER_INCLUDE_DIR
- no changes to core_add_library

finally: flatbuffers is not available for several linux distros, incl ubuntu

**Review by Rechi — CHANGES_REQUESTED** [review]
flatc and flatbuffers has to be added to https://github.com/xbmc/kodi-deps/, so it can be easily build from source (for  windows) if needed.

**Comment by garbear** [comment]
> move all flatbuffer building to depends

I think this is the case. `BuildFlatBuffers.cmake` is not used.

> fix the findmodule to allow bulding with external and internal flatbuffer and just use FLATBUFFER_INCLUDE_DIR

Is there an existing example I can learn from?

> no changes to core_add_library

How do I make other code depend on the generated headers?

**Comment by Rechi** [comment]
FFmpeg, Crossguid, Fmt, fstrcmp and RapidJSON can be built internal.

File: cmake/modules/FindFlatBuffers.cmake [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        COMMENT "Building C++ header for ${FILE}"
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+    endforeach()
+    set(${Name}_OUTPUTS ${FLATC_OUTPUTS} PARENT_SCOPE)
+  endfunction()
+
+  set(FLATBUFFERS_INCLUDE_DIRS ${FLATBUFFERS_INCLUDE_DIR})
+  include_directories(${CMAKE_BINARY_DIR})
+else()
+  set(FLATBUFFERS_INCLUDE_DIR)
```
Comment (Rechi) [inline cmake/modules/FindFlatBuffers.cmake]: If it isn't found you don't have to care about that.

File: cmake/modules/FindFlatBuffers.cmake [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+endif()
+
+include("${FLATBUFFERS_CMAKE_DIR}/BuildFlatBuffers.cmake")
```
Comment (Rechi) [inline cmake/modules/FindFlatBuffers.cmake]: move into `if(FLATBUFFERS_FOUND)` block and add newline at end of file

File: project/BuildDependencies/scripts/0_package.native-win32.list [hunk 4]
```diff
@@ -11,3 +11,4 @@ jsonschemabuilder-1.0.0-win32-3.7z
 swig-3.0.10-win32.7z
 texturepacker-1.1.1-win32.7z
 vswhere-2.2.11-win32.7z
+flatc-1.9.0-win32.7z
```
Comment (Rechi) [inline project/BuildDependencies/scripts/0_package.native-win32.list]: `;PLEASE KEEP THIS LIST IN ALPHABETICAL ORDER!`

File: tools/depends/native/flatbuffers-native/Makefile [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+PREFIX=$(NATIVEPREFIX)
+PLATFORM=$(NATIVEPLATFORM)
+DEPS=../../Makefile.include Makefile
+
+# lib name, version
+LIBNAME=flatbuffers
+VERSION=1.9.0
+SOURCE=v$(VERSION)
+ARCHIVE=$(SOURCE).tar.gz
+BASE_URL=https://github.com/google/flatbuffers/archive
```
Comment (Rechi) [inline tools/depends/native/flatbuffers-native/Makefile]: Has to be uploaded to the mirrors.

File: tools/depends/target/flatbuffers/Makefile [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+VERSION=1.9.0
+SOURCE=v$(VERSION)
+ARCHIVE=$(SOURCE).tar.gz
+
+ifeq ($(PLATFORM),)
+	# Building stand-alone
+	ROOT_DIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
+	PLATFORM = native
+	TARBALLS_LOCATION = $(ROOT_DIR)
+	BASE_URL := https://github.com/google/flatbuffers/archive
```
Comment (Rechi) [inline tools/depends/target/flatbuffers/Makefile]: Has to be uploaded to the mirrors.

File: xbmc/cores/RetroPlayer/savestates/SavestateFlatBuffer.h [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+ *  along with this Program; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+#include "ISavestate.h"
+
+#include "flatbuffers/flatbuffers.h"
```
Comment (Rechi) [inline xbmc/cores/RetroPlayer/savestates/SavestateFlatBuffer.h]: `<flatbuffers/flatbuffers.h>`

File: xbmc/cores/RetroPlayer/messages/CMakeLists.txt (hunk 1 above)
Comment (Rechi) [reply xbmc/cores/RetroPlayer/messages/CMakeLists.txt]: No, but haven't checked how to do it.

File: cmake/modules/FindFlatBuffers.cmake [hunk 8]
```diff
@@ -16,44 +16,82 @@
...
+function(FLATBUFFERS_GENERATE_C_HEADERS Name)
+  set(FLATC_OUTPUTS)
+  foreach(FILE ${ARGN})
+    get_filename_component(FLATC_OUTPUT ${FILE} NAME_WE)
+    set(FLATC_OUTPUT
+      "${CMAKE_CURRENT_BINARY_DIR}/${FLATC_OUTPUT}_generated.h")
+    list(APPEND FLATC_OUTPUTS ${FLATC_OUTPUT})
+
+    add_custom_command(OUTPUT ${FLATC_OUTPUT}
+      COMMAND ${FLATBUFFERS_FLATC_EXECUTABLE}
```
Comment (garbear) [inline cmake/modules/FindFlatBuffers.cmake]: @Rechi when this macro is run, `FLATBUFFERS_FLATC_EXECUTABLE` is empty. How can I propagate its value to this function?

File: cmake/modules/FindFlatBuffers.cmake [hunk 9]
```diff
@@ -16,24 +16,63 @@
...
+  find_program(FLATBUFFERS_FLATC_EXECUTABLE NAMES flatc)
+  find_path(FLATBUFFERS_INCLUDE_DIR NAMES flatbuffers/flatbuffers.h)
+endif()
 
 include(FindPackageHandleStandardArgs)
 find_package_handle_standard_args(flatbuffers
   DEFAULT_MSG FLATBUFFERS_FLATC_EXECUTABLE FLATBUFFERS_INCLUDE_DIR)
 
-if(FLATBUFFERS_FOUND)
   function(FLATBUFFERS_GENERATE_C_HEADERS Name)
```
Comment (garbear) [inline cmake/modules/FindFlatBuffers.cmake]: @Rechi This macro uses `${FLATBUFFERS_FLATC_EXECUTABLE}`, which is defined in this cmake module, but the variable is empty when the macro is run. How can I propagate its value to this macro? It worked before I enabled building internal.

Comment (wsnipex) [reply cmake/modules/FindFlatBuffers.cmake]: it would work if you made it a macro, functions have locally scoped vars.
I'd move the function to Macros.cmake though.

**Comment by wsnipex** [comment]
@garbear see https://github.com/wsnipex/xbmc/commit/67e6b8d6db69dcf421297b4b11a0d7f1ea10fb50 for some cmake fixes.
Not yet perfect at all, but work for building with internal flatbuffers on linux(without depends)

File: tools/depends/native/flatbuffers-native/Makefile (hunk 5 above)
Comment (garbear) [reply tools/depends/native/flatbuffers-native/Makefile]: Can you upload https://github.com/google/flatbuffers/archive/v1.9.0.tar.gz (as `flatbuffers-1.9.0.tar.gz`) to the mirrors?

File: cmake/scripts/common/Macros.cmake [hunk 10]
```diff
@@ -813,3 +813,4 @@ macro(find_addon_xml_in_files)
   # Append also versions.h to depends
   list(APPEND ADDON_XML_DEPENDS "${CORE_SOURCE_DIR}/xbmc/addons/kodi-addon-dev-kit/include/kodi/versions.h")
 endmacro()
+
```
Comment (Rechi) [inline cmake/scripts/common/Macros.cmake]: unrelated

File: tools/depends/target/flatbuffers/Makefile [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	DEPS += ../../Makefile.include
+	BUILD_FLATC=ON
+endif
+
+CMAKE_OPTIONS := -DCMAKE_BUILD_TYPE=Release \
+                 -DFLATBUFFERS_CODE_COVERAGE=OFF \
+                 -DFLATBUFFERS_BUILD_TESTS=OFF \
+                 -DFLATBUFFERS_INSTALL=ON \
+                 -DFLATBUFFERS_BUILD_FLATLIB=OFF \
+                 -DFLATBUFFERS_BUILD_FLATC=$(BUILD_FLATC) \
```
Comment (Rechi) [inline tools/depends/target/flatbuffers/Makefile]: `BUILD_FLATC` is the same for stand alone and depends builds

**Comment by wsnipex** [comment]
small fix for external flatbuffers: https://github.com/wsnipex/xbmc/commit/46527b86db329fdae03607960fa9b8def3b98df7

File: cmake/modules/FindFlatBuffers.cmake [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+                                 -DFLATBUFFERS_INSTALL=ON
+                                 -DFLATBUFFERS_BUILD_FLATLIB=OFF
+                                 -DFLATBUFFERS_BUILD_FLATC=ON
+                                 -DFLATBUFFERS_BUILD_FLATHASH=OFF
+                                 -DFLATBUFFERS_BUILD_GRPCTEST=OFF
+                                 -DFLATBUFFERS_BUILD_SHAREDLIB=OFF
+                                 "${EXTRA_ARGS}"
+                      BUILD_BYPRODUCTS ${FLATBUFFERS_FLATC_EXECUTABLE})
+  set_target_properties(flatbuffers PROPERTIES FOLDER "External Projects"
+                                    INTERFACE_INCLUDE_DIRECTORIES ${FLATBUFFERS_INCLUDE_DIR} ${FLATC_OUTPUTS})
```
Comment (garbear) [inline cmake/modules/FindFlatBuffers.cmake]: `FLATC_OUTPUTS` is defined nowhere

File: cmake/modules/FindFlatBuffers.cmake [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+  set(FLATBUFFERS_FLATC_EXECUTABLE ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/bin/flatc CACHE INTERNAL "FlatBuffer compiler")
+  set(FLATBUFFERS_INCLUDE_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/include CACHE INTERNAL "FlatBuffer include dir")
+  set(FLATBUFFERS_MESSAGES_INCLUDE_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/cores/RetroPlayer/messages CACHE INTERNAL "Generated FlatBuffer headers")
+
+  externalproject_add(flatbuffers
+                      URL ${FLATBUFFERS_URL}
+                      DOWNLOAD_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/download
+                      PREFIX ${CORE_BUILD_DIR}/flatbuffers
+                      CMAKE_ARGS -DCMAKE_INSTALL_PREFIX=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}
+                                 -DCMAKE_BUILD_TYPE=Release
```
Comment (garbear) [inline cmake/modules/FindFlatBuffers.cmake]: @wsnipex @Rechi Debug build fails with internal flatbuffers. `flatc.exe` is not installed for debug builds because debug symbols don't make sense for the compiler, so we need to force it to build in Release mode always.

CMake is ignoring this flag when I select "Debug" in Visual Studio. Is there a way to override cmake's override?

Comment (Rechi) [reply cmake/modules/FindFlatBuffers.cmake]: `flatc.exe`, so you try to use the internal build for Windows? Why?

Comment (garbear) [reply cmake/modules/FindFlatBuffers.cmake]: I assumed we wanted correctness for both modes when building internal. Is this not the case?

**Comment by garbear** [comment]
Added to kodi-deps: https://github.com/xbmc/kodi-deps/pull/6

File: cmake/modules/FindFlatBuffers.cmake (hunk 13 above)
Comment (garbear) [reply cmake/modules/FindFlatBuffers.cmake]: so it's ok that we can't build debug with internal flatbuffers?

Comment (Rechi) [reply cmake/modules/FindFlatBuffers.cmake]: Wasn't I clear enough, why would you want to use internal flatbuffers on windows if you have to use all other precompiled dependencies anyway?

File: cmake/scripts/common/Macros.cmake [hunk 14]
```diff
@@ -76,7 +77,7 @@ function(core_add_library name)
     if(NOT CORE_SYSTEM_NAME STREQUAL windowsstore)
       list(APPEND lib_DEPS dvdnav)
     endif()
-    add_dependencies(${name} ${lib_DEPS})
+    add_dependencies(${name} ${lib_DEPS} ${DEPENDS})
```
[the remark this answers is not in the document]
Comment (garbear) [reply cmake/scripts/common/Macros.cmake]: When I revert this the cmake step fails:

```
CMake Error at xbmc/cores/RetroPlayer/savestates/CMakeLists.txt:16 (add_dependencies):
  Cannot add target-level dependencies to non-existent target
  "retroplayer_savestates".

  The add_dependencies works for top-level logical targets created by the
  add_executable, add_library, or add_custom_target commands.  If you want to
  add file-level dependencies see the DEPENDS option of the add_custom_target
  and add_custom_command commands.
```

I think this is because macros.cmake adds the library, but the savestate CMakeLists.txt tried to add the dependency, which can't see the library target from the macro.

[the remark this answers is not in the document]
Comment (Rechi) [reply cmake/scripts/common/Macros.cmake]: In VS (and XCode) there is only one target (xbmc), where everything is added from the `core_add_library` calls. So the one you want to set as dependency doesn't exist.

[the remark this answers is not in the document]
Comment (garbear) [reply cmake/scripts/common/Macros.cmake]: I see. Does the latest fix look OK?

File: cmake/modules/FindFlatBuffers.cmake (hunk 13 above)
Comment (garbear) [reply cmake/modules/FindFlatBuffers.cmake]: Thanks for explaining

**Comment by garbear** [comment]
jenkins build this please

@wsnipex does 9d4e304 look ok?

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

File: xbmc/cores/RetroPlayer/messages/CMakeLists.txt [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+                     DEPENDS ${_file}
+                     COMMENT "Building C++ header for ${_file}"
+                     WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+endforeach()
+
+add_custom_target(retroplayer_messages DEPENDS ${FLATC_OUTPUTS})
+set_target_properties(retroplayer_messages PROPERTIES FOLDER "Generated Messages"
+                                                      INCLUDE_DIRECTORIES ${CMAKE_CURRENT_BINARY_DIR}
+                                                      SOURCES ${FLATC_OUTPUTS})
+add_dependencies(retroplayer_messages flatbuffers)
```
Comment (Rechi) [inline xbmc/cores/RetroPlayer/messages/CMakeLists.txt]: shouldn't be needed

Comment (garbear) [reply xbmc/cores/RetroPlayer/messages/CMakeLists.txt]: ok, i'll remove

File: cmake/modules/FindFlatBuffers.cmake [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+                                 -DFLATBUFFERS_BUILD_SHAREDLIB=OFF
+                                 "${EXTRA_ARGS}"
+                      PATCH_COMMAND patch -p1 < ${CORE_SOURCE_DIR}/tools/depends/native/flatbuffers-native/0001-Fix-compiler-warning.patch
+                      BUILD_BYPRODUCTS ${FLATBUFFERS_FLATC_EXECUTABLE})
+  set_target_properties(flatbuffers PROPERTIES FOLDER "External Projects"
+                                    INTERFACE_INCLUDE_DIRECTORIES ${FLATBUFFERS_INCLUDE_DIR})
+else()
+  find_program(FLATBUFFERS_FLATC_EXECUTABLE NAMES flatc)
+  find_path(FLATBUFFERS_INCLUDE_DIR NAMES flatbuffers/flatbuffers.h)
+  set(FLATBUFFERS_MESSAGES_INCLUDE_DIR ${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/cores/RetroPlayer/messages CACHE INTERNAL "Generated Flatbuffer headers")
```
Comment (wsnipex) [inline cmake/modules/FindFlatBuffers.cmake]: Please include the 2nd fix I posted earlier instead of doubling this line from above

**Review by wsnipex — COMMENTED** [review]
only one minor, otherwise looks good to me

File: cmake/modules/FindFlatBuffers.cmake (hunk 16 above)
Comment (garbear) [reply cmake/modules/FindFlatBuffers.cmake]: I thought I grabbed the fix you posted? What should this line be?

**Comment by garbear** [comment]
@wsnipex updated. jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
all green

**Comment by garbear** [comment]
jenkins build this please

**Comment by Rechi** [comment]
Fedora has a flatbuffers package, so documentation has to be adjusted.

**Review by wsnipex — APPROVED** [review]

**Comment by garbear** [comment]
added package to fedora. jenkins build this please

**Comment by garbear** [comment]
thanks @Rechi @wsnipex for all your help getting this in!
